src/Components/Web.JS/src/Rendering/StreamingRendering.ts

```diff
@@ -64,20 +67,23 @@ class BlazorStreamingUpdate extends HTMLElement {
 function insertStreamingContentIntoDocument(componentIdAsString: string, docFrag: DocumentFragment): void {
   const markers = findStreamingMarkers(componentIdAsString);
   if (markers) {
+    const { startMarker, endMarker } = markers;
     if (enableDomPreservation) {
-      synchronizeDomContent({ startExclusive: markers.startMarker, endExclusive: markers.endMarker }, docFrag);
+      synchronizeDomContent({ startExclusive: startMarker, endExclusive: endMarker }, docFrag);
     } else {
       // In this mode we completely delete the old content before inserting the new content
-      const { startMarker, endMarker } = markers;
+      const destinationRoot = endMarker.parentNode!;
       const existingContent = new Range();
       existingContent.setStart(startMarker, startMarker.textContent!.length);
       existingContent.setEnd(endMarker, 0);
       existingContent.deleteContents();
 
       while (docFrag.childNodes[0]) {
-        endMarker.parentNode!.insertBefore(docFrag.childNodes[0], endMarker);
+        destinationRoot.insertBefore(docFrag.childNodes[0], endMarker);
       }
     }
+
+    navigationEnhancementCallbacks.documentUpdated();
```

Comment:
This doesn't seem ideal for performance (it's rescanning the entire document after each `<blazor-ssr>` element is added), and I'm not sure it's right for correctness either.

Don't we need to wait until streaming SSR is entirely finished before we start activating interactive components? The prerendered content may include ComponentState generated during an async loading process. The developer won't want interactive components to light up until they have that state, as the whole point of transferring this state is so that during interactive rendering the component doesn't show up in a loading state or duplicate the cost of loading the same data again.

Apologies if I'm misinterpreting what's going on.

Reply:
Yeah, I might not have done a good job picking a descriptive name for that callback. It actually doesn't re-scan the document for `<blazor-ssr>` elements; its purpose is to notify the `RootComponentManager` (or whoever else is interested, in theory) that the document has changed. In the case of `RootComponentManager`, this callback causes it to check each of the component descriptors it already tracks to see if they have been updated or removed (and, if a stream render is not underway, create interactive components for each descriptor that hasn't yet been handled).

> Don't we need to wait until streaming SSR is entirely finished before we start activating interactive components?

Yes, and the current implementation does that. But I do think it would be useful to update parameters for (or remove) already-interactive components. The callback is just a way to signal to whoever's interested that the document has changed and some root-components related operations might need to occur. The actual scanning for component comments happens in the DOM sync algorithm for incoming content, so by the time the content gets sync'd with the document, the `RootComponentManager` is already aware of any new SSR'd components.